daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala

```diff
@@ -328,10 +337,11 @@ class VariableMap private(vTable: Map[GlobalQName, ArrayBuffer[VariableInstance]
         // Evaluate defineVariable statements with non-constant default value expressions
         case (VariableDefined, DataValue.NoValue, true) => {
           val res = inst.defaultValueExpr.get.evaluate(state)
-          inst.setValue(DataValue.unsafeFromAnyRef(res))
+          inst.setDefaultValue(DataValue.unsafeFromAnyRef(res))
```

Comment:
So VariableDefined with value=NoValue is end result of a defineVariable statement without a default value expression.  In this instance, the variable is defined (hence the defineVariable statement) but without a value.  I believe VariableUndefined is only used/checked when resetting a variable we check to make sure it is not undefined as that would mean we are trying to reset a variable that has already been reset.

Reply:
That feels like a bug to me, but maybe I just don't understand what the different states mean (we need to document them). My understanding of the states is:

<dl>
<dt>VariableUndefined</dt>
<dd>The variable has no value, a variable read is an error, setVariable is allowed</dd>
<dt>VariableBeingDefined</dt>
<dd>We are in the process of evaluting and setting the value to a default value, used for circularity detection of defineVariable defaulValue's</dd>
<dt>VariableDefined</dt>
<dd>The variable has a value (either default or externally set), a variable read of that value is allowed, set variable is allowed</dd>
<dt>VariableSet</dt>
<dd>the variable has a value that has been set via setVariable, variable read is allowed, setVariable is no longer allowed</dd>
<dt>VariableRead</dt>
<dd>the variable has been read with a value, that value is now set, variable read is allowed, setVariable is no longer allowed</dd>
<dt>VariableInProcess</dt>
<dd>A suspension to set a variable (via either defaultValue or setVariable?) during unparsing has been created. This variable does not yet have a value. All reads/sets of this variable should block until this state changes.
</dl>

This is my understanding, but it sounds like that's not the case?

Note that another comment I made suggested a change to so that VariableInstance's aren't initialize with NoValue, maybe that would affect some of this?

Reply:
Honestly, that is probably what the different states should be (only VariableDefined/Undefined are different than what you have). I don't think I've changed the current usage patterns of any of these states (other than adding VariableBeingDefined) while I've been making changes to variables, but it would probably be good to add clear documentation on what each state means and make the change to VariableDefined/Undefined.  I don't think it will take much and shouldn't break anything.

I'll take a look at it tomorrow and try to add it to this commit, assuming it doesn't break things or require a large refactor.

Reply:
After taking another look at this, I think we may both have been confused here.  This case of VariableDefined with DataValue.NoValue is used specifically when the defineVariable has a non-constant default value expression.  The value is temporarily set to DataValue.NoValue until forceExpressionEvaluation actually evaluates the non-constant expression and sets the correct value. A more clear solution to this may be to have variables that have a non-constant default value expression remain Undefined until the expression has been evaluated. I'm testing how that change will look now.

Reply:
Please capture the comment above, clarifying the VariableDefined with DataValue.NoValue - what that means and how used, in code comments please. I suggest a comment is needed where VariableDefined is introduced.